src/server/project.ts

```diff
@@ -1773,6 +1774,39 @@ namespace ts.server {
         getIncompleteCompletionsCache() {
             return this.projectService.getIncompleteCompletionsCache();
         }
+
+        /*@internal*/
+        withNoDtsResolutionProject(rootFileNames: string[], cb: (project: Project, ensureRoot: (fileName: string) => boolean) => void) {
+            const options: CompilerOptions = {
+                ...this.getCompilerOptions(),
+                noDtsResolution: true,
+                allowJs: true,
+                maxNodeModuleJsDepth: 3,
+                diagnostics: false,
+                skipLibCheck: true,
+                sourceMap: false,
+                types: ts.emptyArray,
+                lib: ts.emptyArray,
+                noLib: true,
+            };
+            const project = new SingleCommandProject(this.projectService, this.documentRegistry, options);
+            for (const fileName of rootFileNames) {
+                project.addRoot(this.getScriptInfo(fileName)!);
+            }
+            cb(
+                project,
+                fileName => {
+                    const info = project.getScriptInfo(fileName);
+                    if (!info) return false;
+                    if (!project.containsScriptInfo(info)) {
+                        project.addRoot(info);
+                        project.updateGraph();
+                    }
+                    return true;
+                }
+            );
+            project.close();

```

Comment:
We can, but I was concerned about increasing the complexity and the potential for mistakes with respect to the project’s lifecycle. It would also add more file watchers for failed lookups, which I’ve currently short-circuited since the project exists only for a moment in a synchronous function. But if you think it’s worth it, I’m open to trying it.